dma: callback to allow for positive statuses on success

fouge opened the pull request:
Make use of positive status values in the DMA callback to pass info to the DMA client after a successful DMA operation. The status integer is dependent to the DMA implementation and might differ on each platform.

Fixes #52417

cc @teburd

teburd commented:
As mentioned I'm ok with this, but I haven't yet had time to review the code base to verify this catches all the status != 0's.

There's also the potential zephyr downstream users might need to update code as the API contract has changed. So this sort of change should stew for a bit with people able to look at it. I'm not sure if this really requires the process laid out for an API change though. @nashif @mbolivar-nordic any thoughts? Do we need to go through the process for expanding the meaning of a return code?

galak commented:
We should make the return code well defined.  So 0 means DMA complete, and 1 means circular water mark complete.

fouge committed:
dma: callback with 2 status codes for successful transfers

Make use of positive status values in the DMA callback to pass
info to the DMA client after a successful DMA operation.
A completed DMA transfer uses the status 0 while a reached
water mark uses the status 1.

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

fouge commented:
@teburd I updated the PR to reflect the previous discussions and made the code comply with the new API.

teburd commented on drivers/dma/dma_dw_common.c:
```diff
@@ -70,7 +70,7 @@ void dw_dma_isr(const struct device *dev)
 			chan_data->dma_blkcallback(dev,
 						   chan_data->blkuser_data,
-						   channel, 0);
+						   channel, 1);
```
I'd like to see the wording around status 1 updated to include something about "block completion" because for DW DMAC that's what this is.

teburd commented on include/zephyr/drivers/dma.h:
```diff
@@ -140,13 +140,15 @@ struct dma_block_config {
+ * @param dev           Pointer to the DMA device calling the callback.
+ * @param user_data     A pointer to some user data or NULL
+ * @param channel       The channel number
+ * @param status        - 0-transfer complete
```
Please add some defines for the status codes so its easier to read the callback lines.

something like

#define DMA_STATUS_XFER 0
#define DMA_STATUS_BLOCK 1

And use those throughout in place of 0 and 1

Please also note that the STATUS_BLOCK may signify a watermark for when circular buffers are in use in the doc comments here.

teburd commented on tests/drivers/dma/chan_blen_transfer/src/test_dma.c:
```diff
@@ -36,7 +36,7 @@ static char rx_data[RX_BUFF_SIZE] = { 0 };
 		      uint32_t id, int error_code)
 {
-	if (error_code == 0) {
+	if (error_code >= 0) {
```
no longer an error code, should be replaced with status

fouge replied:
I also changed the message printed to print out the status

teburd commented on tests/drivers/dma/chan_link_transfer/src/test_dma.c:
```diff
@@ -42,7 +42,7 @@ static char rx_data2[RX_BUFF_SIZE] = { 0 };
 		      int error_code)
 {
-	if (error_code == 0) {
+	if (error_code >= 0) {
```
replace with status

teburd commented on tests/drivers/dma/loop_transfer/src/test_dma_loop.c:
```diff
@@ -99,7 +99,7 @@ static void dma_user_callback(const struct device *dma_dev, void *arg,
 	}
 
-	zassert_false(error_code, "DMA could not proceed, an error occurred\n");
+	zassert_false(error_code < 0, "DMA could not proceed, an error occurred\n");
```
status

teburd commented on tests/boards/altera_max10/msgdma/src/dma.c:
```diff
@@ -29,7 +29,7 @@ static struct dma_block_config dma_block_cfg = {0};
 			      uint32_t id, int error_code)
 {
-	if (error_code == 0) {
+	if (error_code >= 0) {
```
status

teburd commented:
Thanks for updating this, I think a few nits here need to be addressed, I need to survey the code base a bit in Zephyr as well and ensure this covers at least all Zephyr code.

I'll send out an email this week notifying the community of the change.

fouge commented:
@teburd thanks for the quick review. I updated it with your comments.
I actually found a few more places where the status wasn't correctly set like in `dma_mcux_edma.c`.
Can you have a look at the API doc again please? I hope this is perfectly clear

teburd commented on drivers/dma/dma_dw_common.c:
```diff
@@ -89,7 +89,7 @@ void dw_dma_isr(const struct device *dev)
 			chan_data->dma_tfrcallback(dev,
 						   chan_data->tfruser_data,
-						   channel, 0);
+						   channel, DMA_STATUS_FULL);
```
The define here doesn't really match up with the status, this is a transfer complete and STATUS_FULL in terminology feels different

teburd commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -120,14 +120,14 @@ static bool data_size_valid(const size_t data_size)
 		/* DMA is no longer busy when there are no remaining TCDs to transfer */
 		data->busy = (handle->tcdPool != NULL) && (handle->tcdUsed > 0);
-		ret = 0;
+		ret = DMA_STATUS_FULL;
```
As noted above, transfer done, as in the list of transfers has been completed not a buffer is full.

teburd commented on include/zephyr/drivers/dma.h:
```diff
@@ -136,17 +136,25 @@ struct dma_block_config {
+ * @param dev           Pointer to the DMA device calling the callback.
+ * @param user_data     A pointer to some user data or NULL
+ * @param channel       The channel number
+ * @param status        - 0-DMA_STATUS_FULL buffer fully consumed
```
STATUS_BLOCK is probably fine and could easily relate to the block of data up to the water mark in the circular buffer cases and directly matches with the dma block config.

STATUS_FULL is sort of trying to lean too heavily on the circular buffer terms here when it could really mean the transfer list is complete for nxp edma or dw dma (atmel's xdmac works in a similar way).

STATUS_FULL just feels like the wrong terminology in those cases and needs a tweak

STATUS_DONE? STATUS_COMPLETE? something along those lines maybe.

fouge replied:
I'll change to `STATUS_COMPLETE`

kasunch commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -120,14 +120,14 @@ static bool data_size_valid(const size_t data_size)
 		/* DMA is no longer busy when there are no remaining TCDs to transfer */
 		data->busy = (handle->tcdPool != NULL) && (handle->tcdUsed > 0);
-		ret = 0;
+		ret = DMA_STATUS_COMPLETE;
```
Shouldn't the `data->busy` be checked before setting `ret` to `DMA_STATUS_COMPLETE` or `DMA_STATUS_BLOCK`?

teburd replied:
This is effectively what was being done before with only the status codes being changed, if the driver has a bug in this regard please file an issue and assign to one of the NXP maintainers (e.g. dleach)

fouge commented:
thanks @teburd for the review. I really need this to start working on the UART

MaureenHelm commented:
@galak please revisit

de-nordic reviewed (approved):
Flash drivers look OK.
